### DISCUSSION TITLE
Batch Replace Glyphs

### DISCUSSION THREAD

**Discussion by connordavenport** [fonttools/fonttools#2133] [Q&A]
Hey everyone! I've been struggling to figure this one out — does anyone know of a way to batch replace glyphs in a TTFont from a source? Thanks in advance!

**Comment by connordavenport** [comment]
```python
from fontTools.ttLib import TTFont, newTable
from fontTools.ttLib.tables._g_l_y_f import Glyph
from fontTools.pens.ttGlyphPen import TTGlyphPen
import os 

source = 'sourceFont.ttf'
dest = 'destinationFont.ttf'

s = TTFont(source)
d = TTFont(dest)

for key in ["hyphen", "endash", "emdash"]:

    glyf = d["glyf"]
    sg = s.getGlyphSet()[key]
    glyf[key] = Glyph()
    pen = TTGlyphPen(None)
    sg.draw(pen)

    glyf[key] = pen.glyph()
    glyf[key].program = sg._glyph.program

    d["hmtx"].metrics[key] = s["hmtx"].metrics[key]

d.save(dest)
```
I figured out this very basic version that works with contour-based TTGlyphs. I'd love to see if people have done this with CFF or components. Pens go way over my head :)

**Reply by anthrotype** [reply]
Instead of passing None to the TTGlyphPen, pass in the source glyphsSet, which will be used automatically to resolve component references

**Reply by connordavenport** [reply]
Ah that makes so much sense. Thanks, Cosimo!!

**Reply by myxzlpltk** [reply]
Is there a way to swap glyph letter "a" with letter "x" inside a single font perfectly? It seems some kerning not aligned correctly.

For example, I have a dictionary called `swaps`
```python
swaps = {'z': 'o', 'g': 'c', 'l': 'z', 'c': 'g', 'o': 'l'}
```

And code for swapping will be
```python
from fontTools import ttLib
import copy

font_path = 'input/times.ttf'
font_output = 'exports/times.ttf'

font = ttLib.TTFont(font_path)
original_font = copy.deepcopy(font)

for key, value in swaps.items():
    font['glyf'][key] = original_font['glyf'][value]  # Swap glyph
    font['hmtx'].metrics[key] = original_font['hmtx'].metrics[value] # Swap metric

font.save(font_output)
```
